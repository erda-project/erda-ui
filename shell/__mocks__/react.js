// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

const react = require('react');
// Resolution for requestAnimationFrame not supported in jest error :
// https://github.com/facebook/react/issues/9102#issuecomment-283873039
global.window = global;
window.addEventListener = () => {};
window.requestAnimationFrame = () => {
  // throw new Error('requestAnimationFrame is not supported in Node');
};

module.exports = react;
