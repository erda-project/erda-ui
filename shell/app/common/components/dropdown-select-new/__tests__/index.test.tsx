// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import DropdownSelect from '..';

interface IOption {
  key: string;
  label: string;
  disabled?: boolean;
  children?: IOption[];
}

describe('DropdownSelect', () => {
  const options: IOption[] = [
    {
      key: 'menu-1',
      label: 'menu-1',
    },
    {
      key: 'menu-2',
      label: 'menu-2',
      children: [
        {
          key: 'menu-2-1',
          label: 'menu-2-1',
        },
        {
          key: 'menu-2-2',
          label: 'menu-2-2',
        },
        {
          key: 'menu-2-3',
          label: 'menu-2-3',
          disabled: true,
        },
      ],
    },
    {
      key: 'menu-3',
      label: 'menu-3',
      disabled: true,
    },
  ];
  const itemSelector = '.erda-dropdown-select-option-item';

  it('should work well in basic usage', async () => {
    const clickMenuFn = jest.fn();
    const result = render(<DropdownSelect options={options} onClickItem={clickMenuFn} />);
    expect(result.container).isExist('.erda-dropdown-select', 1);
    fireEvent.click(result.getByText('please select'));
    await waitFor(() => expect(screen.getByRole('menu')).toBeInTheDocument());
    expect(result.baseElement).isExist('.erda-dropdown-select-option-group', 1);

    expect(result.baseElement).isExist(itemSelector, 5);
    fireEvent.click(screen.getByText(options[0].label).closest(itemSelector)!);
    expect(clickMenuFn).toHaveBeenCalledTimes(1);
    expect(clickMenuFn.mock.calls[0]).toEqual([options[0].key, options[0]]);
    // click disabled
    fireEvent.click(screen.getByText('menu-3').closest(itemSelector)!);
    expect(clickMenuFn).toHaveBeenCalledTimes(1);
    // click nest
    fireEvent.click(screen.getByText('menu-2-1').closest(itemSelector)!);
    expect(clickMenuFn).toHaveBeenCalledTimes(2);
    // click nest disabled
    fireEvent.click(screen.getByText('menu-2-3').closest(itemSelector)!);
    expect(clickMenuFn).toHaveBeenCalledTimes(2);
  });

  it('support title, simple mode, size, showFilter, customTrigger', async () => {
    const result = render(
      <DropdownSelect
        options={options}
        // value={options[0].key}
        title="list title"
        mode="simple"
        optionSize="small"
        showFilter
      >
        <div id="trigger">please select</div>
      </DropdownSelect>,
    );
    expect(result.container).isExist('#trigger', 1);
    fireEvent.click(result.container.querySelector('#trigger')!);
    await waitFor(() => expect(screen.getByRole('menu')).toBeInTheDocument());
    expect(screen.getByText('list title')).toBeTruthy();

    // FIXME: chosen item not exist when pass custom children
    // expect(result.baseElement).isExist('.selected-item', 1);

    // TODO: add simple mode test

    expect(result.baseElement).isExist('.small', 5);

    const filterInput = result.baseElement.querySelector('input');
    expect(filterInput).toBeTruthy();

    fireEvent.change(filterInput!, { target: { value: 'menu-2' } });
    expect(result.baseElement).isExist('.erda-dropdown-select-menu input', 1);
    // TODO: how to trigger render?
    // expect(result.baseElement).isExist(itemSelector, 3);
  });
});
