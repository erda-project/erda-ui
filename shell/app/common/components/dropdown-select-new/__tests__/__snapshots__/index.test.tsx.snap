// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownSelect should work well in basic usage 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <script />
    <div>
      <div
        class="ant-dropdown-trigger inline-flex items-center cursor-pointer erda-dropdown-select-content  ant-dropdown-open erda-dropdown-select"
      >
        <div>
          please select
        </div>
      </div>
    </div>
    <div
      style="position: absolute; top: 0px; left: 0px; width: 100%;"
    >
      <div>
        <div
          class="ant-dropdown"
          style="opacity: 0; pointer-events: none;"
        >
          <ul
            class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light erda-dropdown-select-menu"
            data-menu-list="true"
            role="menu"
            tabindex="0"
          >
            <li
              class="ant-dropdown-menu-item erda-dropdown-option-content block"
              data-menu-id="rc-menu-uuid-test-option"
              role="menuitem"
              tabindex="-1"
            >
              <span
                class="ant-dropdown-menu-title-content"
              >
                <div
                  class=" middle cursor-pointer erda-dropdown-select-option-item flex w-full"
                >
                  <div
                    class="flex items-center flex-1 overflow-hidden"
                  >
                    <div
                      class="flex-1 overflow-hidden"
                    >
                      <div
                        class="flex items-center"
                      >
                        <div
                          class="truncate option-label"
                        >
                          menu-1
                        </div>
                      </div>
                    </div>
                  </div>
                  <span
                    class="flex"
                  />
                </div>
                <div
                  class="erda-dropdown-select-option-group w-full"
                >
                  <div
                    class="erda-dropdown-select-option-group-label flex items-center justify-between"
                  >
                    <div
                      class="flex items-center text-xs p-1"
                    >
                      menu-2
                    </div>
                  </div>
                  <div
                    class="erda-dropdown-select-option-group-content"
                  >
                    <div
                      class=" middle cursor-pointer erda-dropdown-select-option-item flex w-full"
                    >
                      <div
                        class="flex items-center flex-1 overflow-hidden"
                      >
                        <div
                          class="flex-1 overflow-hidden"
                        >
                          <div
                            class="flex items-center"
                          >
                            <div
                              class="truncate option-label"
                            >
                              menu-2-1
                            </div>
                          </div>
                        </div>
                      </div>
                      <span
                        class="flex"
                      />
                    </div>
                    <div
                      class=" middle cursor-pointer erda-dropdown-select-option-item flex w-full"
                    >
                      <div
                        class="flex items-center flex-1 overflow-hidden"
                      >
                        <div
                          class="flex-1 overflow-hidden"
                        >
                          <div
                            class="flex items-center"
                          >
                            <div
                              class="truncate option-label"
                            >
                              menu-2-2
                            </div>
                          </div>
                        </div>
                      </div>
                      <span
                        class="flex"
                      />
                    </div>
                    <div
                      class=" middle cursor-pointer erda-dropdown-select-option-item flex w-full"
                    >
                      <div
                        class="flex items-center flex-1 overflow-hidden"
                      >
                        <div
                          class="flex-1 overflow-hidden"
                        >
                          <div
                            class="flex items-center"
                          >
                            <div
                              class="truncate option-label"
                            >
                              menu-2-3
                            </div>
                          </div>
                        </div>
                      </div>
                      <span
                        class="flex"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class=" middle cursor-pointer erda-dropdown-select-option-item flex w-full"
                >
                  <div
                    class="flex items-center flex-1 overflow-hidden"
                  >
                    <div
                      class="flex-1 overflow-hidden"
                    >
                      <div
                        class="flex items-center"
                      >
                        <div
                          class="truncate option-label"
                        >
                          menu-3
                        </div>
                      </div>
                    </div>
                  </div>
                  <span
                    class="flex"
                  />
                </div>
              </span>
            </li>
          </ul>
          <div
            aria-hidden="true"
            style="display: none;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ant-dropdown-trigger inline-flex items-center cursor-pointer erda-dropdown-select-content  ant-dropdown-open erda-dropdown-select"
    >
      <div>
        please select
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
